report/data_sources/hubspot/client.py

```diff
@@ -127,6 +152,60 @@ def get_contacts(self, fields: Iterable[Field]) -> Generator:
 
         yield from self._get_objects(self.api_client.crm.contacts, fields)
 
+    def get_contacts_by_email(self, emails, fields: Iterable[Field]):
+        # We can't have more than 100 items in an IN query, we also can't have
+        # more than 3000 chars in our total request.
+        for email_batch in chunk_with_max_len(emails, chunk_size=100, max_chars=2000):
+            yield from self._get_contacts_by_email(
+                email_batch=email_batch, fields=fields, limit=100
+            )
+
+    def _get_contacts_by_email(
+        self, email_batch, fields: Iterable[Field], limit, max_retries=5
+    ):
+        # Pick the largest size we are allowed
+        limit = max(limit, 100)
+
+        # Example of searching here:
+        # https://github.com/HubSpot/sample-apps-search-results-iterating/blob/main/python/cli.py
+        # That demonstrates pagination, which we are hoping we can get away
+        # without because we ask for 100 emails max.
+        request = PublicObjectSearchRequest(
+            limit=limit,
+            properties=[field.hs_field for field in fields],
+            filter_groups=[
+                FilterGroup(
+                    filters=[
+                        Filter(
+                            property_name="email",
+                            operator="IN",
+                            values=list(email_batch),
+                        )
+                    ]
+                )
+            ],
+        )
+
+        # Example of the retry mechanism here:
+        # https://github.com/HubSpot/sample-apps-rate-limit/blob/master/python/cli.py
+        retries = 0
+        while retries < max_retries:
+            try:
+                response = self.api_client.crm.contacts.search_api.do_search(request)
+                break
+            except ApiException as e:
+                if e.status == HTTPStatus.TOO_MANY_REQUESTS:
+                    print("Rate limit exceeded, retrying in 2 seconds...")
+                    time.sleep(2)
+                    retries += 1
+                else:
+                    raise  # Reraise the exception if it's not a rate limit error
+
+        if len(response.results) == limit:
+            print("Potential pagination error! We got the same size as our batch size")
```

Comment:
It sure would be nice if the API client they provide could:

 * Provide a less weird interface
 * Help us with retries or query rate
 * Smooth over pagination without us having to implement it